Allow to specify HTB burst by duration in ms required to send one burst of data

moeller0 opened the pull request:
This should make it more obvious how much latency under load increase
htb's burst might cause (while probably allowing to use the configured
bandwidth fuller than HTB with the minimal burst). Currently this defaults to
a minimum of 2 MTUs and has no upper bound. The hope is that this makes it simpler to 
specify an acceptable burst parameter.

moeller0 commented:
Hi Toke,
I am trying something new here, instead of pushing half-baked stuff into master I try for a change to work in a branch and allow better review...
The main goal here is to make it easier to play with HTB's burst paramter, but I nte of this works there are two more things worth looking at:
a) instead of taking the unadorned MTU, add overhead and apply ATM/AAL5 expansion to better reflect the size HTB accounts for a normal full-MTU frame
b) set htb_quantum to burst/number of priority tiers, to allow HTB to potentially dequeue from each priority tier, or set quantum = burst.

tohojo commented:
moeller0 <notifications@github.com> writes:

> Hi Toke,
> I am trying something new here, instead of pushing half-baked stuff
> into master I try for a change to work in a branch and allow better
> review...

Excellent idea ;)

My main comment is why keep the old function? Do we really need the
added complexity of supporting both?

> The main goal here is to make it easier to play with HTB's burst
> paramter, but I nte of this works there are two more things worth
> looking at: a) instead of taking the unadorned MTU, add overhead and
> apply ATM/AAL5 expansion to better reflect the size HTB accounts for a
> normal full-MTU frame

Hmm, that's probably over-engineering a bit? :)

> b) set htb_quantum to burst/number of priority tiers, to allow HTB to
> potentially dequeue from each priority tier, or set quantum = burst.

In most cases we won't have backlogs on all tiers simultaneously, will
we? So there's a risk that lowering the quantum will undo the
performance gains of the higher burst size?

-Toke

moeller0 commented:
> My main comment is why keep the old function? Do we really need the
> added complexity of supporting both?

Well, since I can only test in my limited set, I would like the ability to tell people in the field to do A/B testing. I hope that after a bit of testing we should be able to retire one of the implementations again.

> Hmm, that's probably over-engineering a bit? :)

Not sure, with ATM/AAL5 a 1500 MTU frame can turn  into  (1500+48-8)/48 = 32.0833333333 -> 33 * 53 = 1749 Byte paket, but if we only allow a burst of 1500 (HTB's default) we actually can not reliably dequeue a full MTU packet. I am probably overlooking something simple here, like the fact that typically HTB will not try to dequeue without some allocated tokens and hence should never need to dequeue a full packet on the burst credit alone?

> In most cases we won't have backlogs on all tiers simultaneously, will
> we? So there's a risk that lowering the quantum will undo the
> performance gains of the higher burst size?

My hope was that in this case HTB would know and not even try, but that indicates that quantum should be equal to burst, no?

tohojo commented:
moeller0 <notifications@github.com> writes:

>> My main comment is why keep the old function? Do we really need the
>> added complexity of supporting both?
>
> Well, since I can only test in my limited set, I would like the
> ability to tell people in the field to do A/B testing. I hope that
> after a bit of testing we should be able to retire one of the
> implementations again.

Shouldn't we just deal with that like any possible regression?

>> Hmm, that's probably over-engineering a bit? :)
>
> Not sure, with ATM/AAL5 a 1500 MTU frame can turn  into
> (1500+48-8)/48 = 32.0833333333 -> 33 * 53 = 1749 Byte paket, but if we
> only allow a burst of 1500 (HTB's default) we actually can not
> reliably dequeue a full MTU packet. I am probably overlooking
> something simple here, like the fact that typically HTB will not try
> to dequeue without some allocated tokens and hence should never need
> to dequeue a full packet on the burst credit alone?

Which is why we set it to a minimum of two MTU? :)

>> In most cases we won't have backlogs on all tiers simultaneously, will
>> we? So there's a risk that lowering the quantum will undo the
>> performance gains of the higher burst size?
>
> My hope was that in this case HTB would know and not even try, but
> that indicates that quantum should be equal to burst, no?

Yeah, setting them equal is probably a good first approximation, at
least...

-Toke

moeller0 commented:
> Shouldn't we just deal with that like any possible regression?

You tell me ;). Honestly, I need to do more tests and really just wanted to distribute the test version to Pete and @dtaht. I then got carried away after pushing the branch and also pushed the create pull request button. Which might have been premature ;)

> Which is why we set it to a minimum of two MTU? :)

Partly true, the other reason for 2 MTU is that recent VDSL2 with G.INP actually tries to send aggregates on the wire that have a maximum around 3048 Bytes, to allow for two full MTU packets, and I fear that we might fall short on that, but I agree that this is less important as 2 MTU solves most of the theoretical lissues.

tohojo commented:
moeller0 <notifications@github.com> writes:

>> Shouldn't we just deal with that like any possible regression?
>
> You tell me ;). Honestly, I need to do more tests and really just
> wanted to distribute the test version to Pete and @dtaht. I then got
> carried away after pushing the branch and also pushed the create pull
> request button. Which might have been premature ;)

I'd say nuke the old version and go to time-based burst sizes
throughout. We can hold off on merging the PR until after you've had
someone else take the branch for a spin, though... :)

dtaht commented:
I was still pursuing my babel issue this week, sorry. monday.

moeller0 commented:
Okay, this passes some (admittedly light) testing on my side, so I followed your recommendation and retired the classic ranged burst calculation method. This still requires others' to confirm that it works  as desired.

moeller0 commented:
> get_burst() no more present in functions.sh
> get_burst_classic() also removed on next commit

Thanks negan07, I hope I fixed it.

moeller0 commented:
@negan07 I rethought the situation and decided to rename get_bust_by_duration to get_burst to maintain compatibility with existing scripts... if you manage to actually test this branch, I would be delighted if you could keep posting your feed-back here.

negan07 commented:
good

the reason of differentiating i/e target settings is mainly to keep settings independent each other and to help tuning very unbalanced DSL lines: anyway, this is an advanced option and the default values is generally accepted in 99,9% cases.
More tests can be made with highest rates near to Gbps where UL/DL differences may become more significant.

moeller0 commented:
Hi Dave,


> On Oct 8, 2018, at 04:22, Dave Täht <notifications@github.com> wrote:
> 
> @dtaht approved this pull request.

	Great, Thamks. I assume that you tested the most recent version with the verify_qdisc clean-ups?

> 
> Is there any way to reliably get the mtu?

	Well, as far as I can tell we are getting the MTU just fine. It is just that TBF seems to call something MTU that really is not the classical MTU (I tend to agree with TBF though, the burst buffer needs to be able to hold the packages the kernel sends, and the kernel pretends packets are 1514 bytes (which from the kernel's perspective is correct)); htb works-around this issue to default to a burst size of ~1600...


https://elixir.bootlin.com/linux/v4.14.74/source/include/net/pkt_sched.h#L132 :
/* Calculate maximal size of packet seen by hard_start_xmit
   routine of this device.
 */
static inline unsigned int psched_mtu(const struct net_device *dev)
{

	
return dev->mtu + dev->hard_header_len;
}

I believe we do not have access to hard-header_len from user-space, so we could just add say 14 + 4 + 4 to allow for double VLAN tags and be done with? (well we should fix this for the actual burst calculations in simplest_tbf.qos as well).
 

> I'm seeing tagged interfaces a lot.

	For TBF I assume? Do you still see an advantage for simplest_tbf versus simplest?

> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

moeller0 commented:
@tohojo you think we are ready for merging this yet?

tohojo commented:
Sure, I'm fine with merging if you think it's ready. Tried commenting to that effect the other day, but guess GitHub ate my comment...

moeller0 commented:
@EricLuehrsen any comment on this proposed simplification of the burst calculations? I note that not having an upper limit enforced by get_burst is on purpose as now there is a way in defaults.sh for the user to effectively set the duration burst should cover, as long as that duration is >= 1 millisecond

moeller0 commented:
@tohojo Thanks! I guess we should give Eric a chance to comment, as I believe he did the original burst work....

EricLuehrsen commented:
Yes. Burst was minimum of two typical MTU to cover all the conditions robustly. Testing cheap MIPS  that seemed to smooth things out (for the CPU). The goal is for **cheap** hardware to ensure enough bursts to prevent sub-1-ms interrupts from having any load. HTB bounces and returns nothing to send. It doesn't have the overhead of walking its flow structure to find the next packet and see if it fits. The only load is counting tokens until its balance sheet is in the black. This also why as the speed increased, the slope was not to ensure 1 whole ms. If the hardware can actually handle the speed, then it can handle shorter interrupt with real work. Bursting more than 16kB could have unintended consequences, and again, if the CPU can handle the speed, then bursting larger isn't helping. this how I cam to just using 1 ms with an on/off option. More time didn't seem to reduce load.

As for quantum, I would caution against that from being too big. Think about sloshing a tank full of water. Slosh too hard and too fast, and you lose control of the tank. Allowing the flow hierarchy to (un)share quickly my improve performance in test case, but may also cause weird imbalance in real cases. It again should start at some minimum (MTU or so) and maybe increase the calculation at about half of burst.

All the complexity is a consequence of HTB being effectively sample based and structured mentally to old-school systems where an interrupt was often more a suggestion of what to do first on the next 1-10 ms tic. This is where cake tastes much better.

moeller0 commented:
> Yes. Burst was minimum of two typical MTU to cover all the conditions robustly.

I reduced that to MTU plus ATM/AAL5 encapsulation, as for very slow links a single packet already blocks the link painfully long, so no need to make increase the pain, especially since I got rid f the low threshold to apply a custom burst to begin with (I did keep setting burst tp min(calculated_burst, ALAL5'eMTU to have a lower bound so that burst actually can help)

> Bursting more than 16kB could have unintended consequences, and again, if the CPU can handle the speed, then bursting larger isn't helping.

But we have Dave's example of HTB being starved on apu2 hardware at gigabit rates, so I believe scaling to burst duration at the configured bandwidth should be theoretically better than having a hard upper threshold, no?
But do you have a test case for unintended consequences of too large bursts? Note, I just prototyped a way of allowing to specify burst down to 0.001 ms, which should make it possible to better restrict burst even for fast links. I need to test this first, but then I will put it here as well for review.

> As for quantum, I would caution against that from being too big. 

Yes, effectively too high quantum will affect the "mixing" of the different priority tiers, hence I proposed to set quantum to burst/number_of_priority_tiers. I assume though the htb will switch tiers if it runs out of bits to send at the current tier even it quantum is not used up.
For the main reason of making shaping possible at high rates with puny hardware quantum+burst should be an acceptable starting point, no? But I am also fine with max(0.5*burst, AAL5'dMTU) ;)

> All the complexity is a consequence of HTB being effectively sample based and structured mentally to old-school systems where an interrupt was often more a suggestion of what to do first on the next 1-10 ms tic.

Yepp, the way I see burst is as a buffer (and internally htb calls burst buffer and cburst cbuffer) that will isolate htb's exact execution times from starving the link, as any TX-OP not taken will reduce the achievable throughput.

Thanks for your input!

moeller0 commented:
So, from my perspective this should be mergeable now. @dtaht does this out of the box now do the right thing on your apu2? It does not seem to do any harm on my wndr3700v2 (but it also does not gain much, my uplink bandwidth is still in the range sqm could shape even with the old burst calculations).

Note, this does not yet touch the quantum calculations, as I believe these to be a) somewhat orthogonal and b) the current quantum calculations actually do size for a 1ms duration equivalent in the reange of ~30000 to 768000 Kbps. (As I might add for most of the range this actually comes out larger than the old burst calculations, which seems odd).

@tohojo if you want the quantum changes to be rolled into this as well so we have closure on the "batch sizing by required service duration" concept, just holler...

moeller0 commented:
> with sh
> local TMP_BURST_US=$( printf %.0f\\n "${TARGET_BURST_MS}e3" )
> not accepting decimals like 2.66 (invalid number) but 2,66

Mmh, I get:
`root@router:~# TMP=2 ; printf %.0f\\n "${TMP}e3"
2000
root@ router:~# TMP=2.6 ; printf %.0f\\n "${TMP}e3"
2600
root@ router:~# TMP=2,6 ; printf %.0f\\n "${TMP}e3"
ash: invalid number '2,6e3'
0
`
on:

`root@nacktmulle:~# cat /etc/os-release 
NAME="OpenWrt"
VERSION="SNAPSHOT"
ID="openwrt"
ID_LIKE="lede openwrt"
PRETTY_NAME="OpenWrt SNAPSHOT"
VERSION_ID="snapshot"
HOME_URL="http://openwrt.org/"
BUG_URL="http://bugs.openwrt.org/"
SUPPORT_URL="http://forum.lede-project.org/"
BUILD_ID="r8247-90bb790fbf"
LEDE_BOARD="ar71xx/generic"
LEDE_ARCH="mips_24kc"
LEDE_TAINTS="no-all busybox"
LEDE_DEVICE_MANUFACTURER="OpenWrt"
LEDE_DEVICE_MANUFACTURER_URL="http://openwrt.org/"
LEDE_DEVICE_PRODUCT="Generic"
LEDE_DEVICE_REVISION="v0"
LEDE_RELEASE="OpenWrt SNAPSHOT r8247-90bb790fbf"`

So that might be locale dependent...

tohojo commented:
moeller0 <notifications@github.com> writes:

> So that might be locale dependent...

If it's locale dependent, that's a really good reason *not* to do things
that way!

moeller0 commented:
Hi Toke,

> On Oct 11, 2018, at 19:53, Toke Høiland-Jørgensen <notifications@github.com> wrote:
> 
> moeller0 <notifications@github.com> writes:
> 
> > So that might be locale dependent...
> 
> If it's locale dependent, that's a really good reason *not* to do things
> that way!

	I agree, see the edit to the post. So how about awk instead?


> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

moeller0 commented:
Okay:

Macos
`
hms-beagle2:~ smoeller$ TMP=2.6 ; printf %.0f\\n "${TMP}e3"

2600

hms-beagle2:~ smoeller$ TMP=2,6 ; printf %.0f\\n "${TMP}e3"

-bash: printf: 2,6e3: invalid number

0

hms-beagle2:~ smoeller$ LC_NUMERIC="de_DE" ;TMP=2,6 ; printf %.0f\\n "${TMP}e3"

2600

hms-beagle2:~ smoeller$ LC_NUMERIC="de_DE" ;TMP=2.6 ; printf %.0f\\n "${TMP}e3"

-bash: printf: 2.6e3: invalid number

0
`

Openwrt:

`root@nacktmulle:~# TMP=2.6 ; printf %.0f\\n "${TMP}e3"

2600

root@nacktmulle:~# TMP=2,6 ; printf %.0f\\n "${TMP}e3"

ash: invalid number '2,6e3'

0

root@nacktmulle:~# LC_ALL="de_DE" ;TMP=2.6 ; printf %.0f\\n "${TMP}e3"

2600

root@nacktmulle:~# LC_ALL="de_DE" ;TMP=2,6 ; printf %.0f\\n "${TMP}e3"

ash: invalid number '2,6e3'

0
`
(same with setting LC_NUMERIC)

I guess this is makes fprintf unsuitable for our use-case. I guess I will switch to microseconds then as that requires no additional dependency on external binaries...

moeller0 commented:
So thanks to all of you, and sorry for stressing your patience, but I made enough changes to the burst calculations and also switched the quantum calculations to be duration based that I feel that this probably is sufficiently buggy to justify a bit more roasting and review before becoming a candidate for merging. So please test away/
@negan07 thanks again, I beliieve I fixed the comment (and the debug output).

guidosarducci commented:
I would really have to NACK this as it stands. The PR adds a huge amount of complexity relative to the problems it purports to fix: half a dozen new configuration knobs, code changes across 6 different files, and functional rewrites to both quantum *and* burst settings simultaneously. How would you do any useful incremental or A/B testing? I suppose that explains the lack of any up-front analysis or testing...

But the bigger problem by the far is that the PR isn't motivated by trying to understand and fix the root problems, but rather the belief that "specifying the burst size by the time required to service it at the desired bandwidth is a better interface" (sic).

@dtaht  summarizes the actual issues and limited data points rather well (e.g. in https://github.com/tohojo/sqm-scripts/issues/71 and elsewhere):

- discrepancies between sub 200Mbit vs higher rates
- varying reports of problem with both `tbf` and `htb`
- known data point of adequate 1Gbit shaping with explicit burst/cburst of 64KB but diminishing returns above
- estimate that burst and cburst should scale roughly as a function of the bytes that can fit into 1ms

I discovered the root causes/fixes for these items months ago while bug-hunting on my own `sqm-scripts` fork, and the solutions are far simpler than in this PR. Reviewing `get_burst()` and the related math reveals:

1. Effectively and by design , `get_burst()` already scales as a (piece-wise) linear "function of the bytes that can fit into 1ms", with a proportionality constant of **0.5**. The bandwidth range is limited from 24Mbit/s to 216Mbit/s, and the burst values saturate outside this (in theory only :frowning_face: ).
2. The upper bandwidth limit keeps `htb` burst from scaling much higher than 200Mbit. Simply extending the BW range up to 1 Gbps while _still keeping the original proportionality constant_, yields @dtaht 's desired 64KB burst. It seems clear to me the original author (who?) had this in mind.
3. A simple but terrible typo in `get_burst()` results in a `tbf` burst discontinuity above 216 Mbit, dropping it down to a useless 1*MTU instead of reaching 64KB at 1Gbit.
4. Finally, there's a further discontinuity below the 24Mbit/s limit, where the burst falls below 2*MTU, causing it's own problems.

All of these bugs and limitations can be fixed by **three lines of changes** which probably address the vast number of reported problems as well as match the few data points of working high-bandwidth burst values. Here's a minimal patch:
```
diff --git a/src/functions.sh b/src/functions.sh
index e837be2..bdda15e 100644
--- a/src/functions.sh
+++ b/src/functions.sh
@@ -507,17 +507,17 @@ htb_quantum_step() {
 get_burst() {
     MTU=$1
     BANDWIDTH=$2
-    BURST=
+    BURST=$(( ${MTU} * 2 ))

     # 10 MTU burst can itself create delay under CPU load.
     # It will need to all wait for a hardware commit.
     # Note the lean mixture at high bandwidths for upper limit.
     BANDWIDTH_L=$(( ${MTU} *  2 * 8 ))
-    BANDWIDTH_H=$(( ${MTU} * 18 * 8 ))
+    BANDWIDTH_H=$(( ${MTU} * 84 * 8 ))


     if [ ${BANDWIDTH} -gt ${BANDWIDTH_H} ] ; then
-        BURST=$(( ${HTB_MTU} * 10 ))
+        BURST=$(( ${MTU} * 43 ))

     elif [ ${BANDWIDTH} -gt ${BANDWIDTH_L} ] ; then
             # Start with 1ms buffer 2x MTU, and lean out the mixture at higher rates

```
I prefer to take a sound, scientific approach to SW maintenance, where complexity is the enemy and tractability your friend. So a choice between a 3-line change and a sprawling PR is no choice at all. By simply fixing the current bugs and limitations first, one is in a better position to assess the _design_ itself and what does/doesn't work in the field, and then use that to drive further design changes in a comprehensible way. But that's just me and my 2 cents, and what I've done in my own fork...

Best regards all and take care.

moeller0 commented:
Thanks Tony for chiming in. Let me try to address a few issues:

> The PR adds a huge amount of complexity...

In total the PR adds two lines of text to sqm-scripts, not a real explosion in complexity, especially given that the PR adds 18 lines of comment and white-space more than it removes, so measured as lines of code the complexity gets reduced. But even besides the pure numerical reduction of LOC, I truely believe that making the buffer scaling dependent on the time required to empty said buffer (or put differently by the amount of time sqm has to refill the buffer without the hardware running dry) is conceptually simpler than the old "magic" scaling code with its discontinuities and all. 
So would you care to expand why the new mode is conceptually more complex than the old one?

> half a dozen new configuration knobs,

I believe that policy should be under the control of the user, our task is to make a reasonable proposal, but allow overrides as we will not be able to predxict all use cases and circumstances. The fact that the old method did not work out of the box for @dtaht and also did not allow easy overrides supports the addition of new default sizes for htb burst/buffer and htb quantum IMNSHO. The fact the fact that we use interfaces in both directions alone accounts for the fact that instead of two toggles we need 4. I admit that adding the non-directional values might be seen as superfluous, but I believe that if this should ever be exposed in the GUI it should be as a unified value to keep complexity low. But if you insist I am willing to remove the two non-directional duration siz=ing parameters.

> code changes across 6 different files, 

Well, yes if you change a function prototype you better change the users as well, no? But note that the old calling convention should still work but foregoing the ability to control the buffer sizing independently for both directions.

> and functional rewrites to both quantum and burst settings simultaneously.

If you look closely again at the actual code you will notice a clear correlation between the sizing problem for both batching aggregates. Also the changes happen in independent commits. 

So far I see the amount of criticality I expect from our last discussions just in reverse, I could do with a bit more meat and less adrenaline ;)


> But the bigger problem by the far is that the PR isn't motivated by trying to understand and fix the root problems, but rather the belief that "specifying the burst size by the time required to service it at the desired bandwidth is a better interface" (sic).

Well the root problem is that the current implementation does not work as expected and lacks a clear connection to the "Raison d'être" of sqm, namely keeping avoidable latency increases under control. So translating the buffer size to the added latency it might introduce seems like a good fit for sqm's mission.

> I discovered the root causes/fixes for these items months ago while bug-hunting on my own 

Yes and I am really unhappy, that you so far did not offer those patches that fix bugs and short coming independent of your big PR. For example I now realize that you fixed the kernel warning that tbf throws when initialized with a buffer of 1 byte in the intial verif_qdisc, I would have absolutely loved if that would have been fixed in master already*. I believe I tried to ask for that but was summarily ignored. So don't be annoyed when bugs are fixed differently by others.

>Effectively and by design , get_burst() already scales as a (piece-wise) linear "function of the bytes that can fit into 1ms", with a proportionality constant of 0.5. The bandwidth range is limited from 24Mbit/s to 216Mbit/s, and the burst values saturate outside this (in theory only ☹️ ).

Yes that matches my calculations as well. Except that it is only true for get_htb_quantum.

>The upper bandwidth limit keeps htb burst from scaling much higher than 200Mbit. Simply extending the BW range up to 1 Gbps while still keeping the original proportionality constant, yields @dtaht 's desired 64KB burst. It seems clear to me the original author (who?) had this in mind.

Except in [https://github.com/tohojo/sqm-scripts/issues/70](url) Dave talks about 96Kb. The original author @EricLuehrsen had exactly in mind what he coded as far as I can tell (just read [),](https://github.com/tohojo/sqm-scripts/pull/73#issuecomment-428404319) namely a solution that automatically did the right thing for the hardware he had available at the time. But as Dave's apu2 shows the assumptions of that code do not hold for different hardware as well. For me this indicates that automatic scaling does not work over the full range of bandwidth and hardware we want to support in sqm-scripts, hence manual overrides in one of our configuration files, defaults.sh instead of defining policy milits somewhere deep in functions.sh.

> A simple but terrible typo in get_burst() results in a tbf burst discontinuity above 216 Mbit, dropping it down to a useless 1*MTU instead of reaching 64KB at 1Gbit.

Well, as the code is written it intended to clamp burst at 15KB for bandwidth above 216 Mbps, it explicitly did not aim for reaching 64KB. What typo by the way?


>Finally, there's a further discontinuity below the 24Mbit/s limit, where the burst falls below 2*MTU, causing it's own problems.

Care to elaborate which problems you see with leaving burst at the default 1*MTU plus some allowance for overhead? IMHO keeping a minimum of two made only sense until I remembered that at low bandwidths queuing two MTUs introduces a considerable latency, while the CPU most likely is not over-extended so a smaller buffer might well be justified, after all burst/buffer only matters if htb/tbf's execution period falls well below the time required to transmit all data in the lower-lever buffers. 


> All of these bugs and limitations can be fixed by three lines of changes which probably address the vast number of reported problems as well as match the few data points of working high-bandwidth burst values. Here's a minimal patch:

Sure, but that does neither fix the following four issues:

a) the scaling is still controlled by magic constants in the functions

b) there still is a discontinuity, as you just shifted it upward

c) there still is no easy over-ride for the user to control this inherently policy decision.

d) it still leaves burst and quantum massively out of register, as burst tops out at 15KB, while quantum tops out at 96KB.

I have not seen any analysis showing that scaling quantum above burst actually improves things. But with the new implantation and user controls it will be easy to accommodate such a relation much more easily than with the old structure.


> I prefer to take a sound, scientific approach to SW maintenance, where complexity is the enemy and tractability your friend. 

Yes, that pretty much sums up the approach I took here. But you might also note my call to users affected with the problem to go and test this proposal. I initially went and implemented the new mode as a configurable option to allow easier A/B testing, but on @tohojo's recommendation removed the old implementations. Not that A/B testing is that hard between two branches, a simple git checkout bursat_by_duration
[upload scripts to DUT; perform test]
git checkout master
[upload scripts to DUT; perform test]
does not seem to onerous to me.

> So a choice between a 3-line change and a sprawling PR is no choice at all.

I am sad that you see this that way, but it also seems indicates that you did not look through the actual commits (warts and all) given that a number of your comments do not seem to target the actual implementation.

> By simply fixing the current bugs and limitations first, one is in a better position to assess the design itself and what does/doesn't work in the field, and then use that to drive further design changes in a comprehensible way.

Sure, in theory I agree, but here we are talking about a conceptually small change to begin with that also gives the user the, admittedly awkward, method to retain the old behavior for a given bandwidth (by setting the control variables in defaults.sh to appropriate values).



*) Not offering actual bug-fixes indepently of you big rework PR, seem to run afoul of your stated methodology of "simply fixing the current bugs and limitations first", but I digress.

I also want to apologize for the way I have handled the issues I saw with your PR, in hindsite it seems obvious that I must have used the completely wrong tone, otherwise I can not explain your rather emotional take-down of a simple fix for a simple issue.

moeller0 commented:
@tohojo I could try to squash some of the patches to create a somewhat nicer history, but given the low frequency nature of changes in sqm-scripts I personally do not believe this to be necessary. Also if you agree with Tony's analysis, I would love to hear your opinion on what changes you would deem required to get this into mergeable state.

tohojo commented:
moeller0 <notifications@github.com> writes:

> @tohojo I could try to squash some of the patches to create a somewhat
> nicer history, but given the low frequency nature of changes in
> sqm-scripts I personally do not believe this to be necessary. Also if
> you agree with Tony's analysis, I would love to hear your opinion on
> what changes you would deem required to get this into mergeable state.

The main thing is that it is hard to revert if it does turn out to
introduce problems. We could revert the whole thing, of course, but that
would also lose the fixes that's mixed in.

However, I also agree with you that the churn and size of sqm-scripts is
small enough that it's not a huge issue, which is why I haven't brought
it up before... So up to you if you want to squash and rebase :)

dtaht commented:
I would merely like it solved.

I see two passionate opinions (which I love to see!). I lean
to towards the continuous option. 

Get it out there. Have a beer together. 

Toke Høiland-Jørgensen <notifications@github.com> writes:

> moeller0 <notifications@github.com> writes:
>
>> @tohojo I could try to squash some of the patches to create a
> somewhat
>> nicer history, but given the low frequency nature of changes in
>> sqm-scripts I personally do not believe this to be necessary. Also
> if
>> you agree with Tony's analysis, I would love to hear your opinion on
>> what changes you would deem required to get this into mergeable
> state.
>
> The main thing is that it is hard to revert if it does turn out to
> introduce problems. We could revert the whole thing, of course, but
> that
> would also lose the fixes that's mixed in.
>
> However, I also agree with you that the churn and size of sqm-scripts
> is
> small enough that it's not a huge issue, which is why I haven't
> brought
> it up before... So up to you if you want to squash and rebase :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

moeller0 commented:
> @dtaht "I would merely like it solved."

Same here, if you could test the most recent version to confirm whether it does the right thing on your apu2 would help tremendously in justifying merging the "buffer-sizing by duration".

>  I lean to towards the continuous option.

I interpret this as a gentle vote for the by duration approach as that will continuously scale with bandwidth (with a floor at one ATM/AAL5'd MTU, as below one MTU buffer does not make much sense if the goal is to never starve the bottleneck link, below the specified bandwidth). Thanks.

> Get it out there. Have a beer together.

Sounds like a decent proposal, especially since I am well responsible for Tony's healthily critic attitude towards my PR.

EricLuehrsen commented:
When I implemented "burst" settings, HTB was left to its default MTU+. The fastest DOCSIS (worst buffer offenders) I had access to was about 200 Mbps so that is where I stopped testing. What I added was the bare minimum burst buffer I could get to improve things, and no more than I could test for. The problem with cheap hardware keeping up with 25+Mbps was the interrupt rate became fast. The context swap and other overhead eats your useful load (something I am too familiar with engine controls such a V8 engine at 7000RPM). _Most_ designs assume to model around a nominal 1 ms task period, so that was the goal concept for the burst fix. 
- At bandwidth with less than 2 MTU per 1 ms, drop to use HTB default MTU+.
- At bandwidth above, use 2 MTU plus 1-MTU per each additional 2-MTU that could be. 
- At bandwidth where that equation gets to about 10 MTU, stop.

With that in mind, there are areas to improve the equation. It may be desirable to remove the MTU chunk rounding. It may be desirable to add a whole MTU for each one that fits (instead of 1/2). The end-point discontinuities could be less clumsy. However, I will strongly recommend sticking to a 1 ms concept rather than variable time size. Either the hardware needs burst or not.

moeller0 commented:
@EricLuehrsen Thank you very much for the detailed design description above!

My number one goal for the re-coding was conceptual simplicity, and hence we agree that the goal is to size the buffer in a way that "smoothes" over the fact that our shaper might not be executed exactly as we would like to for the purpose of avoiding to under-shape (by not using "tx-ops" that theoretically would be available for the set shaper bandwidth).
So for me it seems obvious that if a buffer should cover a certain duration, that when sizing that buffer we should specify that desired duration explicitly.
I am happy to set the default to 1ms, but keeping the limit hardcoded in a function is IMHO hostile to our users as it assumes that we found the optimal solution, even though this PR indicates that our solution left something to desire... This decidedly is a policy question where the user/administrator needs to decide what trade-off between bandwidth efficiency and latency-under-load-increase they are comfortable with.

I also note that the old code was deviating from the 1ms goal in three ways:

1) MTU quantization, this works sort of okay in a world where all packets are of size MTU but there are a number of smaller packet types (DNS, RTC. VoIP,...) so the whole MTU quantization does not make too much sense, but will basically allow for buffer misssizing by 1*MTU-1. Now this is a rather minor point as I am sure that the differences between quantized and continous will be in the noise, but it certainly is a deviation from the 1ms goal

2) the increase 1 MTU for every 2 MTU in bandwidth will guarantee that that our buffer will be smaller than 1ms for most of the bandwidth range, especially for the higher bandwidths where we expect CPU limits to come up, it seems counter intuitive to use less buffering for higher loads. Now if we would scale the buffer duration by CPU capability, using smaller buffers on beefier CPU, I would agree, but we are not doing that.

3) The whole scaling has non-intuitive dependence on MTU being 1500 Bytes. Changing the MTU will change the buffer size for a given bandwidth:
`((((24000 - 24000) / 16) / 1500) * 1500) + (1500*2)  = 3000 Byte; 3000B*8bpB / 24000kbps = 1ms
((((24000 - 24000) / 16) / 9000) * 9000) + (9000*2) = 18000 Byte; 18000B*8bpB / 24000kbps = 6ms
`
Or rather it would if the code would actually accept these MTUs (but that is an implementation issue, not a conceptual one.

All in all I am of the opinion, that the old code was great for its day and helped sqm scale above the 50-70 Mbps range, but with the prospect of > 1Gbps bandwidths it just does not cut it anymore.

Now, if there are strong opinions on there being a ceiling to the buffer sizing , I would propose to add another two parameters to defaults.sh: EMAX_BUST_B and IMAX_BURST_B and basically add:

```
if [ ${BURST} -gt $[E|I]MAX_BURST_B ] ; then
        BURST=$[E|I]MAX_BURST_B
fi
```

Which in my simple mind is a less opaque interface than hardcoding this in a functions implementation. Now, I am sure I have not won over all hearts and minds here, but from my perspective all this needs is confirmation that it works as designed.

I believe that this PR tackles the issue with a more straight forward implementation that is easier to follow and that has less unexpected dependency on MTU size and will scale well beyond the bandwidths I expect in the intermediate future.

@tohojo what is your opinion on adding a control for the maximum burst-buffer size, and on changing the default from 3ms to 1ms?

moeller0 commented:
> Hmm, I do believe there's a point of diminishing returns on the burst
size, right? 

Yes I believe so, but I have no intuition or data where exactly this cut-off point is, making putting in a hard threshold somewhat tricky. 

> Not sure if it should be exposed as another config knob, though; if users want to
change the size of the burst, there's already a knob to do that?

Mmh, I still think that putting such magic constants into defaults.sh makes more sense then hiding them in the code, unless we know the correct value (like the lower thresholds being 1*MTU plus overhead). I certainly do not envision to expose any of this to the GUI unless there is data showing that this is absolutely necessary. In the spirit of trying to do the right thing by default but still allowing the unexpected (by having the users edit defaults.sh)
Sure you can set the limit (within bounds) by selecting the matching number of microseconds for the BURST_DUR_US parameters, but that seems klunky and will need to be adjusted if bamdwidth is changed. That said, I actually believe that we should not do this right now and see whether we actually need this, but I am prepared to implant and test it if helps convincing people that the PR is an actual improvement...

> And sure, go with 1ms as default...

Will do...

negan07 commented:
https://github.com/tohojo/sqm-scripts/blob/5a9d8f1d8b74bc71f030e6509fb98729d5fa2366/src/functions.sh#L387

on `get_burst()` the failsafe value should be changed to 1000 too

moeller0 commented:
> [sqm-scripts/src/functions.sh](https://github.com/tohojo/sqm-scripts/blob/5a9d8f1d8b74bc71f030e6509fb98729d5fa2366/src/functions.sh#L387)
> Line 387 in [5a9d8f1](/tohojo/sqm-scripts/commit/5a9d8f1d8b74bc71f030e6509fb98729d5fa2366)
>  local SHAPER_BURST_US=3000	# the duration of the burst in microseconds 
> on `get_burst()` the failsafe value should be changed to 1000 too

Ah, good catch, especially for legacy uses like in the original simplest_tbf.qos 1000 us are a better default... Thanks a lot.

So, from what I can tell all we are waiting for is confirmation that this actually does the right thing of Gbps-class of links and routers and we should be able to merge this, no?

@tohojo : How about we wait one more week and merge next weekend if no test results (or bugs) show up one way or the other? Also this time I would like to do without squashing, but promise the next time around to tackle this; all in all I would say using a branch and pull request was an improvement for the quality of implementation and analysis of the issue thanks to everybody adding their input (so thanks, @negan07, @dtaht, @EricLuehrsen, @guidosarducci, and @tohojo).

dtaht commented:
I've had no time for a while. I promise, really, really,
to get that APU2 running again. I stole the PS for something else...

moeller0 <notifications@github.com> writes:

>     sqm-scripts/src/functions.sh
>     Line 387 in 5a9d8f1
>     local SHAPER_BURST_US=3000 # the duration of the burst in
>     microseconds
>     on get_burst() the failsafe value should be changed to 1000 too
>
> Ah, good catch, especially for legacy uses like in the original
> simplest_tbf.qos 1000 us are a better default... Thanks a lot.
>
> So, from what I can tell all we are waiting for is confirmation that
> this actually does the right thing of Gbps-class of links and routers
> and we should be able to merge this, no?
>
> @tohojo : How about we wait one more week and merge next weekend if no
> test results (or bugs) show up one way or the other? Also this time I
> would like to do without squashing, but promise the next time around
> to tackle this; all in all I would say using a branch and pull request
> was an improvement for the quality of implementation and analysis of
> the issue thanks to everybody adding their input (so thanks, @negan07,
> @dtaht, @EricLuehrsen, @guidosarducci, and @tohojo).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

moeller0 commented:
Hi Dave,


> On Oct 22, 2018, at 19:16, Dave Täht <notifications@github.com> wrote:
> 
> 
> I've had no time for a while. I promise, really, really,
> to get that APU2 running again. I stole the PS for something else...

	No hurry, as far as I can tell this now should work better than the the old code (not by being conceptually better, but just by not having the >216Mbps bug) And due to the duration being editable in defaults.sh, users should be able to get back to around the old 15KB limit, simply by setting the duration appropriately (e.g. @1Gbps, 120 us result in a 15KB buffer: 1000000 Kbps * 120us / 8000  = 15000B). By default the 1ms duration will give you 125KB @ 1 Gbps, which seems close to the 64 to 96 KB you tried successfully so far, so I really do not expect any big surprises...
	I really just want this to go one way or the other, but soon ;) 


> 
> moeller0 <notifications@github.com> writes:
> 
> > sqm-scripts/src/functions.sh
> > Line 387 in 5a9d8f1
> > local SHAPER_BURST_US=3000 # the duration of the burst in
> > microseconds
> > on get_burst() the failsafe value should be changed to 1000 too
> >
> > Ah, good catch, especially for legacy uses like in the original
> > simplest_tbf.qos 1000 us are a better default... Thanks a lot.
> >
> > So, from what I can tell all we are waiting for is confirmation that
> > this actually does the right thing of Gbps-class of links and routers
> > and we should be able to merge this, no?
> >
> > @tohojo : How about we wait one more week and merge next weekend if no
> > test results (or bugs) show up one way or the other? Also this time I
> > would like to do without squashing, but promise the next time around
> > to tackle this; all in all I would say using a branch and pull request
> > was an improvement for the quality of implementation and analysis of
> > the issue thanks to everybody adding their input (so thanks, @negan07,
> > @dtaht, @EricLuehrsen, @guidosarducci, and @tohojo).
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub, or mute the thread.
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

yutayu commented:
Hi , I am newbie .
I want to test this pull request, how to ?
My sqm-script shows htb burst 1600bytes always.
sad :P

moeller0 commented:
> Hi , I am newbie .
> I want to test this pull request, how to ?
> My sqm-script shows htb burst 1600bytes always.
> sad :P

The easiest way I can think of is (on a linux computer inside your network):
open a terminal, cd into a diectory where the sqm-scripts source repository should live
`
git clone https://github.com/tohojo/sqm-scripts
cd ./sqm-scripts
git checkout burst_by_duration
cd ./src
scp ./defaults.sh root@192.168.1.1:/usr/lib/sqm/defaults.sh
scp ./functions.sh root@192.168.1.1:/usr/lib/sqm/functions.sh
scp ./simple.qos root@192.168.1.1:/usr/lib/sqm/simple.qos
scp ./simplest.qos root@192.168.1.1:/usr/lib/sqm/simplest.qos
scp ./simplest_tbf.qos root@192.168.1.1:/usr/lib/sqm/simplest_tbf.qos
`

That should get you there; I assume that you have not cloned the sqm-s ripts repository already and that you want to test it on an openwrt router with the IP-address 192.168.1.1.

Question, what bandwidth will you use this with?

yutayu commented:
I am using sqm-script under linux(raspbian)

I saw this. I did it maybe.
https://github.com/TeamPorcupine/ProjectPorcupine/wiki/How-to-Test-a-Pull-Request

and got something error

pi@raspberrypi:~/sqm-scripts $ sudo sqm reload eth0
Stopping SQM on eth0
Starting SQM script: simple.qos on eth0, in: 10500 Kbps, out: 810 Kbps
get_burst (by duration): the calculated burst/quantum size of 101 bytes was below the minimum of 1749 bytes.
get_burst (by duration): the calculated burst/quantum size of 101 bytes was below the minimum of 1749 bytes.
get_burst (by duration): the calculated burst/quantum size of 1312 bytes was below the minimum of 1749 bytes.
get_burst (by duration): the calculated burst/quantum size of 1312 bytes was below the minimum of 1749 bytes.
simple.qos was started on eth0 successfully
pi@raspberrypi:~/sqm-scripts $

yutayu commented:
ty :)

yutayu commented:
pi@raspberrypi:~/sqm-scripts $ tc class show dev eth0
class htb 1:11 parent 1:1 leaf 110: prio rate 128Kbit ceil 270Kbit burst 1600b cburst 1599b 
class htb 1:1 root rate 810Kbit ceil 810Kbit burst 1748b cburst 1748b 
class htb 1:13 parent 1:1 leaf 130: prio rate 135Kbit ceil 794Kbit burst 1748b cburst 1748b 
class htb 1:12 parent 1:1 leaf 120: prio rate 135Kbit ceil 794Kbit burst 1748b cburst 1748b 
class fq_codel 120:2e2 parent 120: 
class fq_codel 110:c0 parent 110: 
class fq_codel 120:2e2 parent 120: 
class fq_codel 110:c0 parent 110:

yutayu commented:
> pi@raspberrypi:~/sqm-scripts $ tc class show dev eth0
> class htb 1:11 parent 1:1 leaf 110: prio rate 128Kbit ceil 270Kbit burst 1600b cburst 1599b
> class htb 1:1 root rate 810Kbit ceil 810Kbit burst 1748b cburst 1748b
> class htb 1:13 parent 1:1 leaf 130: prio rate 135Kbit ceil 794Kbit burst 1748b cburst 1748b
> class htb 1:12 parent 1:1 leaf 120: prio rate 135Kbit ceil 794Kbit burst 1748b cburst 1748b
> class fq_codel 120:2e2 parent 120:
> class fq_codel 110:c0 parent 110:
> class fq_codel 120:2e2 parent 120:
> class fq_codel 110:c0 parent 110:

usually , settled htb burst size is -1. 
for example, sqm-script sets 1749, but actual size is 1748 like this.

yutayu commented:
MTU = 1500
linklayer = atm and stab settings there. 
pppoa line via ethernet.

yutayu commented:
pi@raspberrypi:~/sqm-scripts $ tc class show dev ifb4eth0
class htb 1:10 parent 1:1 leaf 110: prio rate 10500Kbit ceil 10500Kbit burst 1746b cburst 1746b 
class htb 1:1 root rate 10500Kbit ceil 10500Kbit burst 1746b cburst 1746b 
class fq_codel 110:23c parent 110:

negan07 commented:
better to post the entire configuration scheme variables at `/run/sqm/ifb4eth0.state`

yutayu commented:
> better to post the entire configuration scheme variables at `/run/sqm/ifb4eth0.state`
how?

negan07 commented:
run the scripts
setup at least the ul/dl band in kbps

then have a look

yutayu commented:
# Uplink and Downlink values are in kbps
UPLINK=810
DOWNLINK=10500

# SQM recipe to use. For more information, see /usr/lib/sqm/*.help
SCRIPT=simple.qos
#SCRIPT=simplest_tbf.qos

# Optional/advanced config

#ENABLED=1
#QDISC=fq_codel

#LLAM=tc_stab
#LINKLAYER=none
LINKLAYER=atm
#OVERHEAD=0
OVERHEAD=-4
#STAB_MTU=2047
STAB_MTU=1430
#STAB_TSIZE=512
STAB_TSIZE=256
#STAB_MPU=0
STAB_MPU=64

#ILIMIT=
#ELIMIT=
#ITARGET=
#ETARGET=

# ECN ingress resp. egress. Values are ECN or NOECN.
#IECN=ECN
#EECN=ECN
EECN=NOECN

# Extra qdisc options ingress resp. egress
#IQDISC_OPTS=""
#EQDISC_OPTS=""

# CoDel target
#TARGET=5ms

#ZERO_DSCP_INGRESS=1
#IGNORE_DSCP_INGRESS=1

I run script with
sqm start eth0

moeller0 commented:
> I am using sqm-script under linux(raspbian)
> 
> I saw this. I did it maybe.
> https://github.com/TeamPorcupine/ProjectPorcupine/wiki/How-to-Test-a-Pull-Request
> 
> and got something error
> 
> $ sudo sqm reload eth0
> Stopping SQM on eth0
> Starting SQM script: simple.qos on eth0, in: 10500 Kbps, out: 810 Kbps
> get_burst (by duration): the calculated burst/quantum size of 101 bytes was below the minimum of 1749 bytes.
> get_burst (by duration): the calculated burst/quantum size of 101 bytes was below the minimum of 1749 bytes.
> get_burst (by duration): the calculated burst/quantum size of 1312 bytes was below the minimum of 1749 bytes.
> get_burst (by duration): the calculated burst/quantum size of 1312 bytes was below the minimum of 1749 bytes.
> simple.qos was started on eth0 successfully

This is not an error, but rather an informative message that a one millisecond buffer would be smaller than ~1 MTU, and the burst buffer should be at least 1 MTU worth in size. This is the new code working as intended. Thanks for testing.

moeller0 commented:
> usually , settled htb burst size is -1.
for example, sqm-script sets 1749, but actual size is 1748 like this.

Yes, but that is HTB doing this. I believe HTB uses this to make sure the root instance has the smallest buffer and hence control over buffer filling. Again that is pretty much as expected.

moeller0 commented:
`#STAB_MTU=2047
STAB_MTU=1430
#STAB_TSIZE=512
STAB_TSIZE=256
#STAB_MPU=0
STAB_MPU=64`

These seems odd. Tc-stab will use STAB_MTU and STAB_TSIZE to generate a lookup table to get the to be accounted sizes for specific packet sizes. There should be no need to to change these two unless you operate with MTU >> 1500 on the shaped link.
STAB_MTU=64 might be okay or might not be, this depends on whether the link uses ethernet frames including the frame check sequence, but you are using PPPoA (ppp over ATM) which does not actually transmit ethernet frames to begin with, so I would guess your STAB_MPU should actually be zero.
But that is from theory only as I have never personally seen/tested a PPPoA-link.

moeller0 commented:
> My sqm-script shows htb burst 1600bytes always.

This is simply because at your bandwidths a one millisecond buffer would be smaller than the sensible minimum of slightly more than 1 MTU (slightly more to allow for variable overhead, and HTB's automatic deduction by one).

yutayu commented:
I got it . I will try test since now.
This script is Great Job.
:)

moeller0 commented:
> If this is normal, it should not be in the log unless debugging is enabled...

I believe this is worth mentioning, since we "claim" to size the buffer for 1ms clearing time, but in reality we do max(1MTU+, 1ms-clearing-time). But this is your call, and we can delegate this from sqm_log to sqm_debug. I would appreciate though if we could leave it as sqm_log until we have a few more test results in, okay?

tohojo commented:
moeller0 <notifications@github.com> writes:

>> If this is normal, it should not be in the log unless debugging is enabled...
>
> I believe this is worth mentioning, since we "claim" to size the
> buffer for 1ms clearing time, but in reality we do max(1MTU+,
> 1ms-clearing-time). But this is your call, and we can delegate this
> from sqm_log to sqm_debug. I would appreciate though if we could leave
> it as sqm_log until we have a few more test results in, okay?

Sure, keep it in while testing. But maybe stop it from repeating itself? :)

-Toke

moeller0 commented:
> Sure, keep it in while testing. But maybe stop it from repeating itself? :)

Thanks, I believe stopping the repetitions is hard as we have multiple invocations with (potentially) different bandwidth and duration settings; but i see that the current output simply will be too confusing in normal operations and will change to sqm_debug ;)

yutayu commented:
$ tc -d -s qdisc
---
see this line
qdisc fq_codel 110: dev ifb4eth0 parent 1:10 limit 1001p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 

quantum is not 300 .

In addition , Please add for desktop user, add priorize port80,443 mode.
Thank you.

moeller0 commented:
Dear yutayu,

thank you very much for your testing and issue reports.

> qdisc fq_codel 110: dev ifb4eth0 parent 1:10 limit 1001p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn

Yes, this is how simple.qos treats its single tier ingress qdisc in case it is instructed to ignore ingress DSCPs  and also how simplest.qos works. What exactly is your issue with this not being 300?

> In addition , Please add for desktop user, add priorize port80,443 mode.

I fear I will have ot decline this request, part if the idea behind sqm-scripts was to keep QoS manageable by mere mortals, and fudging with port numbers and IP addresses and what ever else other more refined QoS schemes use is opposed to that desired simplicity.
That said, simple.qos offers three priority tiers so you could "simply" mark all packets that you consider to be of higher priority as EF and those that are of background priority as CS1/BK and you should get most of what you seem to desire.

One more thing, and I am unsure how to say this with all the politeness I intend, could you please post these issues and observations that are not directly related to this PR's change of burst and quantum calculations to be defined by time as new issues [()](url
https://github.com/tohojo/sqm-scripts/issues). This will make it less likely they get burried at the end of an already longish message thread.

Anyways, thanks for your time.

yutayu commented:
It seems network stops sometimes.
master is good.

yutayu commented:
I think you are right . moe.
but something is still wrong this branch 

1:11 and ifb4eth0 have differenent burst sizes.

tc class show dev eth0
class htb 1:11 parent 1:1 leaf 110: prio rate 128Kbit ceil 271Kbit burst 1600b cburst 1599b 
class htb 1:1 root rate 815Kbit ceil 815Kbit burst 1748b cburst 1748b 
class htb 1:13 parent 1:1 leaf 130: prio rate 135Kbit ceil 799Kbit burst 1748b cburst 1748b 
class htb 1:12 parent 1:1 leaf 120: prio rate 135Kbit ceil 799Kbit burst 1748b cburst 1748b 
class fq_codel 120:1f7 parent 120: 
class fq_codel 110:374 parent 110: 
class fq_codel 120:1f7 parent 120: 
class fq_codel 110:374 parent 110: 
tc class show dev ifb4eth0
class htb 1:10 parent 1:1 leaf 110: prio rate 11603Kbit ceil 11603Kbit burst 1746b cburst 1746b 
class htb 1:1 root rate 11603Kbit ceil 

I

moeller0 commented:
> 1:11 and ifb4eth0 have differenent burst sizes.

Yes, this is how simple.qos operates ATM it dooes not set a burst buffer for 1:11, but that is an orthogonal issue to this patchset, as that existed already before this branch came to life.

@dtaht do you remember whether there is a specific reason for not setting burst on 1:11 or is that a simple oversight that we might remedy while we are at it.

moeller0 commented:
Thanks everybody for your input and testing. I believe the proposed solution is better forit! 
Now this branch got a bit messy, so I squashed it down to three commits and offered these as a new PR. I would like to re-direct any further discussion to the new PR (#78) and close this.
